src/lualib/Error.ts

```diff
@@ -0,0 +1,56 @@
+interface ErrorType<T> extends Function {
+    name: string;
+    new (...args: any[]): T;
+}
+
+function __TS__GetErrorStack(constructor: Function): string {
+    let level = 1;
+    while (true) {
+        const info = debug.getinfo(level, "f");
+        level += 1;
+        if (!info) {
+            // constructor is not in call stack
+            level = 1;
+            break;
+        } else if (info.func === constructor) {
+            break;
+        }
+    }
+    return debug.traceback(undefined, level);
+}
+
+function __TS__GetErrorString(this: void, error: Error): string {
+    return error.message !== "" ? `${error.name}: ${error.message}` : error.name;
+}
+
+function __TS__InitErrorClass<T>(Type: ErrorType<T>, name?: string): any {
+    if (name) {
+        Type.name = name;
+    }
+    return setmetatable(Type, {
+        __call: (_self: any, message: string) => new Type(message),
+    });
+}
+
+Error = __TS__InitErrorClass(
+    class Error {
+        public name = "Error";
+        public message: string;
+        public stack: string;
+
+        constructor(message = "") {
+            this.message = message;
+            this.stack = __TS__GetErrorStack((this.constructor as any).new);
+            getmetatable(this).__tostring = __TS__GetErrorString;
```

Comment:
I wanted to suggest `metatable.__tostring = metatable.__tostring || __TS__GetErrorString`, but actually that would break inheritance:

```ts
class CustomError { toString() { return '...' } }
class ExtendedCustomError extends CustomError {}
```

Not sure how that would affect performance, but to support this case we'd have to walk through prototype chain or reconsider doing it during transformation.

Reply:
Actually I was thinking about adding `metatable.__tostring = metatable.__tostring || __TS__GetErrorString`. Why would that break inheritance?

Reply:
Because `ExtendedCustomError` prototype wouldn't have `__tostring` I think? I haven't tested it though, so if it works that's fine. We can add this as a test though

Reply:
It works, because it's `__index` metatable has `__tostring`. tbh though I would rather remove this solution as we are currently over-engineering the error class to work around a more general issue of tstl's class system (see #728).

Reply:
Actually, we'll have to hack it in any case, since user's `toString` implementation should have only description part. We'll have to wrap it with something like that:

```ts
function __TS__WrapErrorToString<T extends Error>(this: void, errorObj: T, getDescription: (this: T) => string): string {
    const description = getDescription.call(errorObj);
    const caller = debug.getinfo(3, "f");
    if (_VERSION === "Lua 5.1" || (caller && caller.func !== error)) {
        return description;
    } else {
        return `${description}\n${errorObj.stack}`;
    }
}

Error = __TS__InitErrorClass(
    class {
        public name = "Error";
        public stack: string;

        constructor(public message = "") {
            this.stack = __TS__GetErrorStack((this.constructor as any).new);
            const metatable = getmetatable(this);
            if (!metatable.__stackPatched) {
                metatable.__stackPatched = true;
                metatable.__tostring = __TS__WrapErrorToString(this, (this as any).__tostring);
            }
        }

        public toString(): string {
            return this.message !== "" ? `${this.name}: ${this.message}` : this.name;
        }
    },
    "Error"
);
```